apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetItemService.java

```diff
@@ -207,6 +224,116 @@ private NotFoundException failWithNotFound(String message) {
                 Response.status(Response.Status.NOT_FOUND).entity(new ErrorMessage(List.of(message))).build());
     }
 
+    /**
+     * Enriches dataset items with trace metadata by fetching trace details and adding them to the item data.
+     * 
+     * @param workspaceId the workspace ID for context
+     * @param items the list of dataset items to enrich (modified in place)
+     * @return a Mono that completes when all items have been enriched
+     */
+    private Mono<Void> enrichItemsWithTraceMetadata(String workspaceId, List<DatasetItem> items) {
+        // Get unique trace IDs from items
+        Set<UUID> traceIds = items.stream()
+                .map(DatasetItem::traceId)
+                .filter(Objects::nonNull)
+                .collect(Collectors.toSet());
+
+        log.debug("Found '{}' unique trace IDs to enrich", traceIds.size());
+
+        if (traceIds.isEmpty()) {
+            log.warn("No trace IDs found in dataset items, skipping metadata enrichment");
+            return Mono.empty();
+        }
+
+        // Fetch trace details for all traces
+        return Flux.fromIterable(traceIds)
+                .flatMap(traceId -> traceService.get(traceId)
+                        .map(trace -> {
+                            log.debug("Successfully fetched trace '{}' with metadata", traceId);
+                            return Map.entry(traceId, trace);
+                        })
+                        .onErrorResume(error -> {
+                            log.warn("Failed to fetch trace details for trace '{}': {}", traceId, error.getMessage());
+                            return Mono.empty();
+                        }))
+                .collectMap(Map.Entry::getKey, Map.Entry::getValue)
+                .doOnNext(traceMap -> {
+                    log.debug("Successfully fetched '{}' traces, enriching dataset items", traceMap.size());
+                    // Enrich items with trace metadata by replacing them in the list
+                    for (int i = 0; i < items.size(); i++) {
+                        DatasetItem item = items.get(i);
+                        if (item.traceId() != null && traceMap.containsKey(item.traceId())) {
+                            Trace trace = traceMap.get(item.traceId());
+                            DatasetItem enrichedItem = enrichItemWithTraceMetadata(item, trace);
+                            log.debug("Enriched dataset item '{}' with trace metadata", item.id());
+                            items.set(i, enrichedItem);
```

Comment:
Modifying the list you are iterating over is dangerous. We don't need to do this; it's better to update the reference since you are already doing it on `enrichItemWithTraceMetadata`.